frame/evm/precompile/bls12377/src/lib.rs

```diff
@@ -0,0 +1,620 @@
+pub struct BLS12377G1Add;
+
+impl BLS12377G1Add {
+	/// https://eips.ethereum.org/EIPS/eip-2539#g1-addition
+	const GAS_COST: u64 = 600;
+}
+
+impl Precompile for BLS12377G1Add {
+	/// Implements EIP-2539 G1Add precompile.
+	/// > G1 addition call expects `256` bytes as an input that is interpreted as byte concatenation of two G1 points (`128` bytes each).
+	/// > Output is an encoding of addition operation result - single G1 point (`128` bytes).
+	fn execute(handle: &mut impl fp_evm::PrecompileHandle) -> PrecompileResult {
+		handle.record_cost(BLS12377G1Add::GAS_COST)?;
+
+		let input = handle.input();
+		if input.len() != 256 {
+			return Err(PrecompileFailure::Error {
+				exit_status: ExitError::Other("Input must contain 256 bytes".into()),
+			});
+		}
+
+		// Decode G1 point p_0
+		let p0 = decode_g1(input, 0)?;
+		// Decode G1 point p_1
+		let p1 = decode_g1(input, 128)?;
+		// Compute r = p_0 + p_1
+		let r = p0 + p1;
+		// Encode the G1 point into 128 bytes output
+		let output = encode_g1(r.into_affine());
+
+		Ok(PrecompileOutput {
+			exit_status: ExitSucceed::Returned,
+			output: output.to_vec(),
+		})
+	}
+}
+
+/// BLS12377G1Mul implements EIP-2539 G1Mul precompile.
+pub struct BLS12377G1Mul;
+
+impl BLS12377G1Mul {
+	/// https://eips.ethereum.org/EIPS/eip-2539#g1-multiplication
+	const GAS_COST: u64 = 12_000;
+}
+
+impl Precompile for BLS12377G1Mul {
+	/// Implements EIP-2539 G1Mul precompile.
+	/// > G1 multiplication call expects `160` bytes as an input that is interpreted as byte concatenation of encoding of G1 point (`128` bytes) and encoding of a scalar value (`32` bytes).
+	/// > Output is an encoding of multiplication operation result - single G1 point (`128` bytes).
+	fn execute(handle: &mut impl fp_evm::PrecompileHandle) -> PrecompileResult {
+		handle.record_cost(BLS12377G1Mul::GAS_COST)?;
+
+		let input = handle.input();
+		if input.len() != 160 {
+			return Err(PrecompileFailure::Error {
+				exit_status: ExitError::Other("Input must contain 160 bytes".into()),
+			});
+		}
+
+		// Decode G1 point
+		let p = decode_g1(input, 0)?;
+		// Decode scalar value
+		let e = decode_fr(input, 128);
+		// Compute r =  = e * p
+		let r = p.mul(e);
+		// Encode the G1 point into 128 bytes output
+		let output = encode_g1(r.into_affine());
+
+		Ok(PrecompileOutput {
+			exit_status: ExitSucceed::Returned,
+			output: output.to_vec(),
+		})
+	}
+}
+
+/// BLS12377G1MultiExp implements EIP-2539 G1MultiExp precompile.
+pub struct BLS12377G1MultiExp;
+
+impl BLS12377G1MultiExp {
+	const MULTIPLIER: u64 = 1_000;
+
+	/// Returns the gas required to execute the pre-compiled contract.
+	fn calculate_gas_cost(input_len: usize) -> u64 {
+		// Calculate G1 point, scalar value pair length
+		let k = input_len / 160;
+		if k == 0 {
+			return 0;
+		}
+		// Lookup discount value for G1 point, scalar value pair length
+		let d_len = BLS12377_MULTIEXP_DISCOUNT_TABLE.len();
+		let discount = if k < d_len {
```

Comment:
> For this case there was a table prepared for discount in case of k <= 128 points in the multiexponentiation with a discount cup max_discount for k > 128.

the final result is the same, would you consider changing this to `k <= d_len`?